### DISCUSSION TITLE
Contextvars propagation to back ground task

### DISCUSSION THREAD

**Discussion by yyamano** [encode/starlette#1975] [Q&A] [answered]
Is it possible to propapate contextvars to a back ground task?

Here is my simple server code.

```
import contextvars
from time import sleep
import uuid

from starlette.applications import Starlette
from starlette.responses import JSONResponse
from starlette.routing import Route
from starlette.background import BackgroundTasks

request_id = contextvars.ContextVar('request_id')
def bar():
    print(f"bar: request_id={request_id.get()}")
    
def foo(request):
    request_id.set(uuid.uuid4())
    print(f"foo: request_id={request_id.get()}")
    tasks = BackgroundTasks()
    tasks.add_task(bar)
    return JSONResponse({'message': 'foo'}, background=tasks)
    
routes = [
    Route('/foo', endpoint=foo, methods=['GET'])
]
app = Starlette(routes=routes)
```

I get the following error when I send a GET request to http://localhost:8000/foo.

```
> poetry run uvicorn buz:app
INFO:     Started server process [22440]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
foo: request_id=c6b3dac2-62f4-4525-a1d3-d9f6bd5c3d69
INFO:     127.0.0.1:49240 - "GET /foo HTTP/1.1" 200 O

[snip]

  File "C:\Users\yyamano\work\buz.py", line 12, in bar
    print(f"bar: request_id={request_id.get()}")
LookupError: <ContextVar name='request_id' at 0x000002238E9AACF0>
```

**Comment by alex-oleshkevich** [comment, marked as answer]
1. Try to make `foo` async function.
2. In this example, you'd better to pass request id directly to the task without using globals.
I personally try to avoid any globals if possible

**Reply by gouthampai05** [reply]
I ran into the same issue and dug into the root cause.

Sync routes run on a threadpool thread via `run_in_threadpool`, but background tasks execute on the event loop thread. `ContextVar` values are thread-scoped and don't cross that boundary, so anything set in a sync handler is silently lost by the time the background task runs. Async routes are unaffected because both run on the same event loop thread.

OpenTelemetry for instance stores trace context in a ContextVar internally. Users have no access to it and no way to pass it explicitly. This means OTel trace context silently drops from background tasks in sync routes, which is a reported issue in the OTel Python instrumentation: open-telemetry/opentelemetry-python-contrib#3586

The fix is to capture the context at schedule time via `contextvars.copy_context()` and run the task inside that snapshot via `ctx.run()`. This fixes propagation for both sync and async cases and follows the same design principle as `asyncio.Task`, which does exactly this so that tasks run in the context of their caller rather than whatever context happens to be active at execution time.

Happy to open a PR with tests if this is the right direction.
